v2/apierror/apierror.go

```diff
@@ -233,30 +233,54 @@ func (a *APIError) Metadata() map[string]string {
 
 }
 
-// FromError parses a Status error or a googleapi.Error and builds an APIError.
-func FromError(err error) (*APIError, bool) {
-	if err == nil {
-		return nil, false
-	}
-
-	ae := APIError{err: err}
+// setDetailsFromError parses a Status error or a googleapi.Error
+// and sets status and details or httpErr and details, respectively.
+// It returns false if neither Status nor googleapi.Error can be parsed.
+func (a *APIError) setDetailsFromError(err error) bool {
 	st, isStatus := status.FromError(err)
 	var herr *googleapi.Error
 	isHTTPErr := errors.As(err, &herr)
 
 	switch {
 	case isStatus:
-		ae.status = st
-		ae.details = parseDetails(st.Details())
+		a.status = st
+		a.details = parseDetails(st.Details())
 	case isHTTPErr:
-		ae.httpErr = herr
-		ae.details = parseHTTPDetails(herr)
+		a.httpErr = herr
+		a.details = parseHTTPDetails(herr)
 	default:
+		return false
+	}
+	return true
+}
+
+// FromError parses a Status error or a googleapi.Error and builds an APIError.
+func FromError(err error) (*APIError, bool) {
+	if err == nil {
 		return nil, false
 	}
 
+	ae := APIError{err: err}
+	if !ae.setDetailsFromError(err) {
+		return nil, false
+	}
 	return &ae, true
+}
+
+// FromWrappingError parses a Status error or a googleapi.Error and
+// builds an APIError, but to avoid a cycle when the provided error
+// will externally wrap the new APIError, it does not set err in the
+// new APIError.
+func FromWrappingError(err error) (*APIError, bool) {
```

Comment:
I am not sold on the name/sig of this function. What if it was something like `ParseError(err error, wrap bool) (*APIError, bool)`? The From error should just call that function with true set?

Reply:
PTAL at the changes to generated code in googleapis/google-api-go-client/pull/1730, and let me know if you're sure you want the extra noise of the boolean in the calls. Keeping those inline calls streamlined was my motivation. Happy to shorten the name to `ParseError` if you like that.

Reply:
Sorry, my mistake, I just remembered `FromWrappingError` is only [called in one location](https://github.com/googleapis/google-api-go-client/blob/3c3aca86c0eda954a1fb67a9e188a9917d480e60/internal/gensupport/error.go) in 1730. I'll update to add the bool.

Reply:
Fixed, thanks!